cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/TypeResolver.kt

```diff
@@ -30,190 +30,37 @@ import de.fraunhofer.aisec.cpg.graph.Component
-     *
-     * @param type newly created Type
-     * @return If the same type was already stored in the typeState Map the stored one is returned.
-     *   In the other case the parameter type is stored into the map and the parameter type is
-     *   returned
-     */
-    protected fun obtainType(type: Type): Type {
-        return if (type.root == type && type in typeState) {
-            typeState.keys.first { it == type }
-        } else {
-            addType(type)
-            type
-        }
-    }
-
-    /**
-     * Responsible for storing new types into typeState
-     *
-     * @param type new type
-     */
-    protected fun addType(type: Type) {
-        if (type.root == type && type !in typeState) {
-            // This is a rootType and is included in the map as key with empty references
-            typeState[type] = mutableListOf()
-            return
-        }
-
-        // ReferencesTypes
-        if (type.root in typeState) {
-            if (type !in (typeState[type.root] ?: listOf())) {
-                typeState[type.root]?.add(type)
-                addType((type as SecondOrderType).elementType)
-            }
-        } else {
-            addType(type.root)
-            addType(type)
-        }
-    }
-
-    protected fun removeDuplicateTypes() {
-        // Remove duplicate firstOrderTypes
-        firstOrderTypes.addAll(typeManager.firstOrderTypes)
-
-        // Propagate new firstOrderTypes into secondOrderTypes
-        val secondOrderTypes = typeManager.secondOrderTypes
-        for (t in secondOrderTypes) {
-            t.root = firstOrderTypes.firstOrNull { it == t.root } ?: t.root
-        }
-
-        // Build Map from firstOrderTypes to list of secondOderTypes
-        for (t in firstOrderTypes) {
-            typeState[t] = mutableListOf()
-        }
-
-        // Remove duplicate secondOrderTypes
-        secondOrderTypes.forEach { processSecondOrderTypes(it) }
-
-        // Remove duplicates from fields
-        secondOrderTypes.forEach { removeDuplicatesInFields(it) }
-    }
-
-    /**
-     * Visits all FirstOrderTypes and replace all the fields like returnVal or parameters for
-     * FunctionPointertype or Generics for ObjectType
-     *
-     * @param t FirstOrderType
-     */
-    protected fun removeDuplicatesInFields(t: Type) {
-        // Remove duplicates from fields
-        if (t is FunctionPointerType) {
-            t.returnType = obtainType(t.returnType)
-            t.parameters = t.parameters.map(::obtainType)
-        } else if (t is ObjectType) {
-            t.generics = t.generics.map(::obtainType)
-        }
-    }
-
     /**
      * Pass on the TypeSystem: Sets RecordDeclaration Relationship from ObjectType to
      * RecordDeclaration
      *
      * @param component
      */
     override fun accept(component: Component) {
-        removeDuplicateTypes()
         val walker = IterativeGraphWalker()
-        walker.registerOnNodeVisit(::ensureUniqueType)
```

Comment:
We still walk through every single node in the graph. Is this intended? Most of them are irrelevant as we only care about `RecordDeclaration`s in the handle function.

Reply:
I will change it to an AST listener only listening for records.